dotnet/src/SemanticKernel.Abstractions/Memory/MemoryModelException.cs

```diff
@@ -0,0 +1,35 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+
+namespace Microsoft.SemanticKernel.Memory;
+
+/// <summary>
+/// Exception thrown when a failure occurs while trying to convert memory models for storage or retrieval.
+/// </summary>
+public class MemoryModelException : KernelException
```

Comment:
I wonder whether the API callers need to distinguish whether the failure is related to model operation or service operation. If not, one exception can be used, and the failure details can be communicated via Message and Data properties unless there is a need to have those details available via exception custom properties.

Reply:
I would say yes, they would want to distinguish. Service call failures are often transient, and can then be retried later, while mapping errors are usually code bugs, which can be record specific, and would never be retriable (at least for that record).  I think if we can add more information to the operation exception about the source of the failure, that would be even better, e.g. does the collection not exist, or is the service down, or do we just not know.  If we do something like that though, we would be able to consolidate the two exceptions into one operation one, where mapping errors are just another error code inside.

Reply:
I would think a little bit more about naming here, since it was hard to understand exception reason from name until I read XML doc. 

First of all, we should probably make naming consistent for this type of entity. For example, if we called `TDataModel` generic type on class level, maybe we can use `DataModel` naming in exception as well. I'm also wondering if you considered keyword `Schema` for this.

Also, we could specify what action caused this error, because with current name it's not clear whether it's initialization exception or something else. 

And last recommendation would be to stick with one naming for `Convert`/`Map` action in methods and XML docs to avoid confusion. If there is `JsonObjectCustomMapper`, then we should probably use `Map/Mapping` in all places. 

Based on above and also previous discussions regarding `Memory` keyword, we can name this exception as `VectorDataModelMappingException`, as example :)

Reply:
Thanks for the suggestions.  I've renamed the exception to MemoryDataModelMappingException for now, and will rename further to match the vector naming when I rename the other classes.  I also want to do some work to see if it won't be possible to remove this exception altogether by just having some error code on the operation exception.  I'll investigate that next.

Reply:
I also like what the Entity Framework folks have done, and I'm thinking that mimicking some of the Exception types available there probably makes sense: https://learn.microsoft.com/en-us/dotnet/api/system.data.entityexception?view=netframework-4.8.1